client/executor/src/wasm_runtime.rs

```diff
@@ -104,86 +124,122 @@ impl RuntimesCache {
 	///
 	/// `Error::InvalidMemoryReference` is returned if no memory export with the
 	/// identifier `memory` can be found in the runtime.
-	pub fn fetch_runtime<E: Externalities>(
-		&mut self,
-		ext: &mut E,
+	pub fn with_instance<'c, R, F>(
+		&self,
+		code: CodeSource<'c>,
+		ext: &mut dyn Externalities,
 		wasm_method: WasmExecutionMethod,
 		default_heap_pages: u64,
 		host_functions: &[&'static dyn Function],
-	) -> Result<(&mut (dyn WasmRuntime + 'static), &RuntimeVersion, Vec<u8>), Error> {
-		let code_hash = ext
-			.original_storage_hash(well_known_keys::CODE)
-			.ok_or(Error::InvalidCode("`CODE` not found in storage.".into()))?;
-
-		let heap_pages = ext
-			.storage(well_known_keys::HEAP_PAGES)
-			.and_then(|pages| u64::decode(&mut &pages[..]).ok())
-			.unwrap_or(default_heap_pages);
+		allow_missing_func_imports: bool,
+		f: F,
+	) -> Result<Result<R, Error>, Error>
+		where F: FnOnce(&dyn WasmInstance, Option<&RuntimeVersion>, &mut dyn Externalities) -> Result<R, Error>,
+	{
+		let (code_hash, heap_pages) = match &code {
+			CodeSource::Externalities => {
+				(
+					ext
+						.original_storage_hash(well_known_keys::CODE)
+						.ok_or(Error::InvalidCode("`CODE` not found in storage.".into()))?,
+					ext
+						.storage(well_known_keys::HEAP_PAGES)
+						.and_then(|pages| u64::decode(&mut &pages[..]).ok())
+						.unwrap_or(default_heap_pages)
+				)
+			},
+			CodeSource::Custom(code) => {
+				(sp_core::blake2_256(code).to_vec(), default_heap_pages)
+			}
+		};
 
-		let result = match self.instances.entry((wasm_method, code_hash.clone())) {
-			Entry::Occupied(o) => {
-				let result = o.into_mut();
-				if let Ok(ref mut cached_runtime) = result {
-					let heap_pages_changed = cached_runtime.heap_pages != heap_pages;
-					let host_functions_changed = cached_runtime.runtime.host_functions()
-						!= host_functions;
-					if heap_pages_changed || host_functions_changed {
-						let changed = if heap_pages_changed {
-							"heap_pages"
-						} else {
-							"host functions"
-						};
+		let runtimes = self.runtimes.upgradable_read(); // this must be released prior to calling f
```

Comment:
Why an upgradable read lock? As far as I can see, we never lock this for reading? 

`upgradable_read()` can only be hold by one thread at a time, like a write lock. So, we can use a write lock directly.

Reply:
Right, I'll replace it with a `Mutex`. It is only used here and unlocked for the actual call anyway.